scripts/validations.py

```diff
@@ -0,0 +1,125 @@
+
+UNLISTED_ADDONS = set()
+
+
+def parse_validations(results):
+    return (json.loads(result) for result in results)
+
+
+def unlisted_validations(results, unlisted_addons=None):
+    if unlisted_addons is None:
+        unlisted_addons = UNLISTED_ADDONS
+    return (result
+            for result in results
+            if (not result['metadata'].get('listed', True)
+                or result['metadata'].get('id') in unlisted_addons))
+
+
+def severe_validations(results):
+    return (result
+            for result in results
+            if (result['signing_summary']['high'] > 0 or
+                result['signing_summary']['medium'] > 0 or
+                result['signing_summary']['low'] > 0))
+
+
+def error_messages(results):
+    return ({'addon': result['metadata'].get('id', 'unknown-addon'),
+             'message_id': '.'.join(message['id']),
+             'context': message['context']}
+            for result in results
+            for message in result['messages']
+            if 'signing_severity' in message)
+
+
+def sort_by_message(results):
+    return sorted(results, key=lambda r: r['message_id'])
+
+
+def group_by_message(results):
+    return itertools.groupby(results, lambda r: r['message_id'])
+
+
+def extract_error_results(results):
+    for error, messages in results:
+        all_messages = list(messages)
+        yield {
+            'error': error,
+            'total': len(all_messages),
+            'unique': len(set(msg['addon'] for msg in all_messages)),
+            'contexts': [msg['context'] for msg in all_messages],
+        }
+
+
+def sort_results_by_unique(results):
+    return sorted(results, reverse=True, key=lambda r: r['unique'])
+
+
+def format_error_count(results):
+    return ('{error} {total} {unique}'.format(**result)
+            for result in results)
+
+
+def format_contexts(results):
+    for result in results:
+        for context in result['contexts']:
+            yield json.dumps({
+                'error': result['error'],
+                'context': context,
+            })
+
+
+def parse_unlisted_addons():
+    global UNLISTED_ADDONS
+    with open('validations/unlisted-addons.txt') as f:
+        UNLISTED_ADDONS = set(guid.strip() for guid in f)
+
+
+def main(action):
+    pipeline = [
+        parse_validations,
+        unlisted_validations,
+        severe_validations,
+        error_messages,
+        sort_by_message,
+        group_by_message,
+        extract_error_results,
+        sort_results_by_unique,
+    ]
+
+    if action == ACTION_CONTEXT:
+        pipeline.append(lambda results: itertools.islice(results, 5))
+        pipeline.append(format_contexts)
+    elif action == ACTION_COUNT:
+        pipeline.append(format_error_count)
+        parse_unlisted_addons()
+    else:
+        raise ValueError('{0} is not a valid action'.format(action))
+
+    # Read from STDIN.
+    val = sys.stdin
+
+    # Process through the pipeline.
+    for fn in pipeline:
+        val = fn(val)
```

Comment:
I wrote this using `reduce` at one point but went back to the loop since this is pretty straightforward and using `reduce` moves all the complexity onto one line.

It looked like this which felt "unbalanced". I think if the inner line was longer I would be okay with it but having a long line in the `for` with something so simple inside of it set off some flags for me.

```python
for line in reduce(lambda val, fn: fn(val), pipeline, sys.stdin):
    print line